### PR TITLE
Added pyodide_main to be able to run onload page

### PR DESCRIPTION
**Pull request by redradist**

### CODE REVIEW

**Comment by rth** [comment]
Thanks @redradist ! I can see the use case for it, and it looks like [bryton](https://github.com/brython-dev/brython) does something like this.

I have several comments/concerns though,
 - Adding support for  `<script type="text/python">` makes sense but it's not an [official MIME type](https://en.wikipedia.org/wiki/Media_type) so technically maybe it should be `text/x.python` to be in the "Unregistered tree" space. Not sure about it.
 - The immediate question when I see this is how do I run some python snippet with some python packages installed. The current way to do it [is convoluted](https://pyodide.readthedocs.io/en/latest/pypi.html#complete-example) and it this could a way to simplify the syntax it could be interesting. For instance to paraphrase the bryton example something like,
```html
        <head>
            <script type="text/javascript" src="/path/to/pyodide.js"></script>
        </head>

        <body onload="pyodide_main(['numpy'])">

            <script type="text/python">
              # some python code
            </script>
        </body>
```

where if the list of requirements is empty it does what is currently done in this PR, and otherwise first loads micropip, then `micropip.install(requirements)`  and finally runs the scripts (as done in the above linked example, except returning a promise to JS as in https://github.com/iodide-project/pyodide/issues/680). I'm not saying we need to implement it in this PR, but we should check that there would be some way to run this script types with a set of packages.

- please add an example of using this at the end of `docs/using_pyodide_from_javascript.md` and also add the added function to `docs/api_reference.md`.

**Comment by rth** [comment]
Let's see what other reviewers think about this. Maybe @wlach ?

**Comment by redradist** [comment]
> 
> 
> Thanks @redradist ! I can see the use case for it, and it looks like [bryton](https://github.com/brython-dev/brython) does something like this.
> 
> I have several comments/concerns though,
> 
>     * Adding support for  `<script type="text/python">` makes sense but it's not an [official MIME type](https://en.wikipedia.org/wiki/Media_type) so technically maybe it should be `text/x.python` to be in the "Unregistered tree" space. Not sure about it.
> 
>     * The immediate question when I see this is how do I run some python snippet with some python packages installed. The current way to do it [is convoluted](https://pyodide.readthedocs.io/en/latest/pypi.html#complete-example) and it this could a way to simplify the syntax it could be interesting. For instance to paraphrase the bryton example something like,
> 
> 
> ```
>         <head>
>             <script type="text/javascript" src="/path/to/pyodide.js"></script>
>         </head>
> 
>         <body onload="pyodide_main(['numpy'])">
> 
>             <script type="text/python">
>               # some python code
>             </script>
>         </body>
> ```
> 
> where if the list of requirements is empty it does what is currently done in this PR, and otherwise first loads micropip, then `micropip.install(requirements)` and finally runs the scripts (as done in the above linked example, except returning a promise to JS as in #680). I'm not saying we need to implement it in this PR, but we should check that there would be some way to run this script types with a set of packages.
> 
>     * please add an example of using this at the end of `docs/using_pyodide_from_javascript.md` and also add the added function to `docs/api_reference.md`.

You are right, it was inspired by `Brython` ;)

Even though "text/python" is not officially supported MIME type, it gives good interoperability with `Brython` and it also nice to have `Python` text in browser )))

**Comment by redradist** [comment]
@rth Could you check the ci/circleci: test-chrome, because I cannot view it ... CircleCI tries to access to all my organizations and at current time I do not want to allow it ;)

Could you send me test console message ?

**Comment by rth** [comment]
There are couple of unrelated test failures; tests are a bit unstable particularly on chrome. You can ignore those. Anyway the added code wouldn't be directly tested by the test suite.

**Comment by redradist** [comment]
> 
> 
> There are couple of unrelated test failures; tests are a bit unstable particularly on chrome. You can ignore those. Anyway the added code wouldn't be directly tested by the test suite.

@rth Thanks ;)

**Comment by phorward** [comment]
Hi @redradist, I like the idea with the `<script>`-Tag you used used here. But, as far as I can see, you're only evaluating the code directly provided inside the script-tags. Why not, additionally, make it even possible to download (and install) code provided by a `src`-attribute?

I already did several iterations over this topic, my first draft was #489 but it now ended up using a pre-defined JSON file as list of modules to be loaded (see [here](https://github.com/mausbrand/pyodide-app-stub/blob/master/app.js)). Your draft using the `<script>`-tag could be a nice alternative for this when src is supported also. I think it's not that much work to do additionally for this. The only important thing to do is save the downloaded files into one location where Pyodide can locate (and import!) the modules correctly.

Do you want to give it a try?

**Comment by redradist** [comment]
@phorward 
> 
> 
> Hi @redradist, I like the idea with the `<script>`-Tag you used used here. But, as far as I can see, you're only evaluating the code directly provided inside the script-tags. Why not, additionally, make it even possible to download (and install) code provided by a `src`-attribute?
> 
> I already did several iterations over this topic, my first draft was #489 but it now ended up using a pre-defined JSON file as list of modules to be loaded (see [here](https://github.com/mausbrand/pyodide-app-stub/blob/master/app.js)). Your draft using the `<script>`-tag could be a nice alternative for this when src is supported also. I think it's not that much work to do additionally for this. The only important thing to do is save the downloaded files into one location where Pyodide can locate (and import!) the modules correctly.
> 
> Do you want to give it a try?

Added downloading `Python` scripts from `script[src]` ;)

File: src/pyodide.js [hunk 1]
```diff
@@ -462,4 +462,40 @@ var languagePluginLoader = new Promise((resolve, reject) => {
...
+    }
+  }
+  return pythonScripts;
+}
+function _downloadPythonScript(url, callback) {
+  var request = new XMLHttpRequest();
+  request.open('GET', url);
+  request.responseType = 'text';
+  request.onload = function() {
+    callback(request.response);
```
Comment (rth) [inline src/pyodide.js:481]: what would happen here if the HTTP status is 404 or 500?

**Comment by phorward** [comment]
> Added downloading `Python` scripts from `script[src]` ;)

Please provide a test suite for this. I'm sure that not all edge-cases are covered with this solution. What about imports, and failing imports?

**Comment by wlach** [comment]
> Let's see what other reviewers think about this. Maybe @wlach ?

Hey, I don't think I have time to evaluate this in detail right now but in principle this sounds useful. I trust your judgement @rth!

**Comment by albertas** [comment]
I am really glad to see this feature being developed. Is there any way I could help to finish this PR?

**Comment by rth** [comment]
Thanks for you interest @albertas ! The main points that would need to be addressed,
 - find a way to test this as part of our test suite
 - have some API for the list of dependecies/packages to load before executing the code snippet.
 - add documentation

see the above discussion for more details.

I'm not sure if @redradist is still planning to work on this PR, if not feel free to open a follow-up PR with those improvements.

**Comment by jeremiahrose** [comment]
What's the status on this pull request / what's required to get it working and merged??

**Comment by rth** [comment]
@jeremiahrose There was no progress since my last comment above, which also includes what's needed to merge it. Feel free to take this PR over and open a new one. The JS package had significant refactorings since, so it might be easier to copy this function manually there rather than solving merge conflicts.

**Comment by rth** [comment]
I would say this use case is now solved by pyscript (which uses pyodide), and so in the end I'm not sure we should go in this direction ourselves. Closing. Please comment if you disagree.
